File: exercises/ddd_infrastructure_layer/typescript/test/infrastructure/JSONFileApplicationRepository.test.ts
import { beforeEach, expect, it } from "bun:test";
import { ApplicationId } from "../../src/domain/application/ApplicationId";
import { Application } from "../../src/domain/application/Application";
import { ApplicationStatus } from "../../src/domain/application/ApplicationStatus";
import { JSONFileApplicationRepository } from "../../src/infrastructure/JSONFileApplicationRepository";
import { buildApplicationWithBasicDetails, buildApplicationWithMotivationLetter } from "../helpers/applicationFixtures";

if (process.env.RUN_INFRA_TESTS) {

  const TEST_FILE = "data/test-applications.json";
  const applicationUuid = "11111111-1111-1111-1111-111111111111";

  function resetTestFile() {
    const initialData = `[]`; // If your store structure is different, adjust this initial data accordingly
    Bun.write(TEST_FILE, initialData);
  }

  // Ensure the test file is reset before each test
  beforeEach(() => {
    resetTestFile();
  });


  it("stores and retrieves an application with basic details", async () => {
    let repo1 = new JSONFileApplicationRepository(TEST_FILE);
    let application = buildApplicationWithBasicDetails(applicationUuid);
    await repo1.save(application);

    // We create a second repository instance to ensure data is actually read from the file
    let repo2 = new JSONFileApplicationRepository(TEST_FILE);
    let retrieved = await repo2.getById(new ApplicationId(applicationUuid));
    expect(retrieved).toBeDefined();
    expect(retrieved?.id.value).toBe(applicationUuid);
    expect(retrieved?.receivedDateTime.value).toEqual(application.receivedDateTime.value);
    expect(retrieved?.fullName?.value).toBe(application.fullName!.value);
    expect(retrieved?.phoneNumber?.value).toBe(application.phoneNumber!.value);
    expect(retrieved?.emailAddress?.value).toBe(application.emailAddress!.value);
  });

  it("returns undefined when retrieving a non-existent application", async () => {
    let repo = new JSONFileApplicationRepository(TEST_FILE);
    let retrieved = await repo.getById(new ApplicationId("00000000-0000-0000-0000-000000000000"));
    expect(retrieved).toBeUndefined();
  });

  it("stores and retrieves an application with all details", async () => {
    let application = buildApplicationWithMotivationLetter(applicationUuid);
    let repo1 = new JSONFileApplicationRepository(TEST_FILE);
    await repo1.save(application);

    // We create a second repository instance to ensure data is actually read from the file
    let repo2 = new JSONFileApplicationRepository(TEST_FILE);
    let retrieved = await repo2.getById(new ApplicationId(applicationUuid));
    expect(retrieved).toBeDefined();
    expect(retrieved?.id.value).toBe(applicationUuid);
    expect(retrieved?.receivedDateTime.value).toEqual(application.receivedDateTime.value);
    expect(retrieved?.fullName?.value).toBe(application.fullName!.value);
    expect(retrieved?.phoneNumber?.value).toBe(application.phoneNumber!.value);
    expect(retrieved?.emailAddress?.value).toBe(application.emailAddress!.value);
    expect(retrieved?.motivationLetter?.value).toBe(application.motivationLetter!.value);
  });

  it("stores and retrieves a rejected application", async () => {
    const application = buildApplicationWithMotivationLetter(applicationUuid);
    application.reject();
    let repo1 = new JSONFileApplicationRepository(TEST_FILE);
    await repo1.save(application);

    // We create a second repository instance to ensure data is actually read from the file
    let repo2 = new JSONFileApplicationRepository(TEST_FILE);
    let retrieved = await repo2.getById(new ApplicationId(applicationUuid));
    expect(retrieved).toBeDefined();
    expect(retrieved?.id.value).toBe(application.id.value);
    expect(retrieved?.receivedDateTime.value).toEqual(application.receivedDateTime.value);
    expect(retrieved?.fullName?.value).toBe(application.fullName!.value);
    expect(retrieved?.phoneNumber?.value).toBe(application.phoneNumber!.value);
    expect(retrieved?.emailAddress?.value).toBe(application.emailAddress!.value);
    expect(retrieved?.motivationLetter?.value).toBe(application.motivationLetter!.value);
    expect(retrieved?.stage).toBe(application.stage);
    expect(retrieved?.status).toBe(ApplicationStatus.Rejected);
  });
}

File: exercises/ddd_infrastructure_layer/java/src/test/java/infrastructure/JSONFileApplicationRepositoryTest.java
package infrastructure;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNull;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.UUID;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;

import domain.application.Application;
import domain.application.ApplicationId;
import domain.application.ApplicationStatus;
import domain.application.stages.ApplicationStages;
import helpers.ApplicationFixtures;

@Tag("infrastructure")
public final class JSONFileApplicationRepositoryTest {
    private static final Path TEST_FILE = Path.of("data", "test-applications.json");
    private static final UUID APPLICATION_ID = UUID.fromString("11111111-1111-4111-1111-111111111111");
    private static final UUID OTHER_APPLICATION_ID = UUID.fromString("99999999-9999-4999-9999-999999999999");

    @BeforeEach
    void resetTestFile() throws Exception {
        Files.createDirectories(TEST_FILE.getParent());
        Files.writeString(TEST_FILE, "[]");
    }

    @Test
    void storesAndRetrievesApplicationWithBasicDetails() throws Exception {
        Application application = ApplicationFixtures.buildEmptyApplication(APPLICATION_ID);
        JSONFileApplicationRepository repo1 = new JSONFileApplicationRepository(TEST_FILE);

        repo1.save(application);

        JSONFileApplicationRepository repo2 = new JSONFileApplicationRepository(TEST_FILE);
        Application result = repo2.getById(application.getId());

        assertEquals(APPLICATION_ID, result.getId().getValue());
        assertEquals(ApplicationFixtures.FIXED_DATE_TIME, result.getReceivedDateTime().getValue());
        assertEquals("Alex Example", result.getFullName().getValue());
        assertEquals("+447700900123", result.getPhoneNumber().getValue());
        assertEquals("alex@example.com", result.getEmailAddress().getValue());
    }

    @Test
    void returnsNullWhenApplicationDoesNotExist() throws Exception {
        JSONFileApplicationRepository repository = new JSONFileApplicationRepository(TEST_FILE);

        Application retrieved = repository.getById(new ApplicationId(OTHER_APPLICATION_ID));

        assertNull(retrieved);
    }

    @Test
    void storesAndRetrievesApplicationWithAllDetails() throws Exception {
        Application application = ApplicationFixtures.buildApplicationWithMotivationLetter(APPLICATION_ID);
        application.approveInitialApplication();
        JSONFileApplicationRepository repo1 = new JSONFileApplicationRepository(TEST_FILE);

        repo1.save(application);

        JSONFileApplicationRepository repo2 = new JSONFileApplicationRepository(TEST_FILE);
        Application result = repo2.getById(application.getId());

        assertEquals(APPLICATION_ID, result.getId().getValue());
        assertEquals(ApplicationFixtures.FIXED_DATE_TIME, result.getReceivedDateTime().getValue());
        assertEquals("Alex Example", result.getFullName().getValue());
        assertEquals("+447700900123", result.getPhoneNumber().getValue());
        assertEquals("alex@example.com", result.getEmailAddress().getValue());
        assertEquals("I am very motivated to join.", result.getMotivationLetter().getValue());
        assertEquals(ApplicationStages.THREE, result.getStage());
        assertEquals(ApplicationStatus.IN_PROGRESS, result.getStatus());
    }

    @Test
    void storesAndRetrievesRejectedApplication() throws Exception {
        Application application = ApplicationFixtures.buildApplicationWithMotivationLetter(APPLICATION_ID);
        application.reject();
        JSONFileApplicationRepository repo1 = new JSONFileApplicationRepository(TEST_FILE);

        repo1.save(application);

        JSONFileApplicationRepository repo2 = new JSONFileApplicationRepository(TEST_FILE);
        Application result = repo2.getById(application.getId());

        assertEquals(APPLICATION_ID, result.getId().getValue());
        assertEquals(ApplicationFixtures.FIXED_DATE_TIME, result.getReceivedDateTime().getValue());
        assertEquals("Alex Example", result.getFullName().getValue());
        assertEquals("+447700900123", result.getPhoneNumber().getValue());
        assertEquals("alex@example.com", result.getEmailAddress().getValue());
        assertEquals("I am very motivated.", result.getMotivationLetter().getValue());
        assertEquals(ApplicationStages.TWO, result.getStage());
        assertEquals(ApplicationStatus.REJECTED, result.getStatus());
    }
}
